server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -121,28 +207,189 @@ public ClusterState closeIndices(ClusterState currentState, final Index[] indice
+                logger.debug("index {} has been blocked before closing but is now deleted, ignoring", index);
+                listener.onResponse(new AcknowledgedResponse(true));
+                return;
+            } else if (indexRoutingTable.allPrimaryShardsUnassigned()) {
+                logger.debug("index {} has been blocked before closing but is now unassigned, ignoring", index);
+                listener.onResponse(new AcknowledgedResponse(true));
+                return;
+            }
+
+            final ImmutableOpenIntMap<IndexShardRoutingTable> shards = indexRoutingTable.getShards();
+            final AtomicArray<AcknowledgedResponse> results = new AtomicArray<>(shards.size());
+            final CountDown countDown = new CountDown(shards.size());
+
+            for (IntObjectCursor<IndexShardRoutingTable> shard : shards) {
+                final IndexShardRoutingTable shardRoutingTable = shard.value;
+                final ShardId shardId = shardRoutingTable.shardId();
+                sendVerifyShardBeforeCloseRequest(shardRoutingTable, timeout, new ActionListener<ReplicationResponse>() {
+                    @Override
+                    public void onResponse(final ReplicationResponse replicationResponse) {
+                        ReplicationResponse.ShardInfo shardInfo = replicationResponse.getShardInfo();
+                        results.setOnce(shardId.id(), new AcknowledgedResponse(shardInfo.getFailed() == 0));
+                        processIfFinished();
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
+                        results.setOnce(shardId.id(), new AcknowledgedResponse(false));
+                        processIfFinished();
+                    }
+
+                    private void processIfFinished() {
+                        if (countDown.countDown()) {
+                            final boolean acknowledged = results.asList().stream().allMatch(AcknowledgedResponse::isAcknowledged);
+                            listener.onResponse(new AcknowledgedResponse(acknowledged));
+                        }
+                    }
+                });
+            }
         }
 
-        //no explicit wait for other nodes needed as we use AckedClusterStateUpdateTask
-        return  allocationService.reroute(
-            ClusterState.builder(updatedState).routingTable(rtBuilder.build()).build(),
-            "indices closed [" + indicesAsString + "]");
+        private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shardRoutingTable, @Nullable final TimeValue timeout,
+                                                       final ActionListener<ReplicationResponse> listener) {
+            final ShardId shardId = shardRoutingTable.shardId();
+            if (shardRoutingTable.primaryShard().unassigned()) {
+                logger.debug("primary shard {} is unassigned, ignoring", shardId);
```

Comment:
maybe log the routing entry (`shardRoutingTable.primaryShard()`), it might contain a little more info, e.g. unassigned reason

Reply:
+1